src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs

```diff
@@ -5,248 +5,300 @@
-        /// Makes an expression that constructs the return value for the deconstruction:
-        /// a series of tuple constructions, that are chained with the help of placeholders.
-        /// The placeholders that are set are added to the list for later clearing.
-        /// </summary>
-        private BoundExpression MakeReturnValue(BoundDeconstructionAssignmentOperator node, ArrayBuilder<BoundValuePlaceholderBase> placeholders)
+        private BoundExpression EvaluateConversionToTemp(BoundExpression expression, Conversion conversion,
+            TypeSymbol destinationType, ArrayBuilder<LocalSymbol> temps, ArrayBuilder<BoundExpression> effects)
         {
-            BoundExpression loweredConstruction = null;
-            foreach (var constructionInfo in node.ConstructionStepsOpt)
+            if (conversion.IsIdentity)
             {
-                // All the input placeholders for the constructions should already be set
-                loweredConstruction = (BoundExpression)Visit(constructionInfo.Construct);
-
-                AddPlaceholderReplacement(constructionInfo.OutputPlaceholder, loweredConstruction);
-                placeholders.Add(constructionInfo.OutputPlaceholder);
+                return expression;
             }
-
-            Debug.Assert(loweredConstruction != null);
-            return loweredConstruction;
+            var evalConversion = MakeConversionNode(expression.Syntax, expression, conversion, destinationType, @checked: false);
+            return EvaluateSideEffectingArgumentToTemp(evalConversion, effects, ref temps);
         }
 
-        /// <summary>
-        /// Adds the side effects to stores and returns temporaries (as a flat list) to access them.
-        /// </summary>
-        private ImmutableArray<BoundExpression> GetAssignmentTargetsAndSideEffects(BoundTupleExpression variables, ArrayBuilder<LocalSymbol> temps, ArrayBuilder<BoundExpression> stores)
+        private ImmutableArray<BoundExpression> InvokeDeconstructMethod(DeconstructionInfo deconstruction, BoundExpression target,
+            ArrayBuilder<BoundExpression> effects, ref ArrayBuilder<LocalSymbol> temps)
         {
-            var assignmentTargets = ArrayBuilder<BoundExpression>.GetInstance(variables.Arguments.Length);
-
-            variables.VisitAllElements(
-                (variable, args) =>
-                {
-                    if (variable.Kind == BoundKind.DiscardExpression)
-                    {
-                        args.targets.Add(variable);
-                    }
-                    else
-                    {
-                        args.targets.Add(args.self.TransformCompoundAssignmentLHS(variable, args.stores, args.temps, isDynamicAssignment: variable.Type.IsDynamic()));
-                    }
-                },
-                (temps: temps, stores: stores, targets: assignmentTargets, self: this)
-            );
+            AddPlaceholderReplacement(deconstruction._inputPlaceholder, target);
```

Comment:
It also seems easy to get into a case where a subtree is lowered or even evaluated more than once. In other similar cases we just pass temps as parameters.
Is  "deconstruction._invocation" too complex to be handled via some kind of `MakeDeconstructionInvocation(..., loweredOutputExpr, inputTemps)`  ?

---
In reply to: [100884679](https://github.com/dotnet/roslyn/pull/17027#discussion_r100884679) [](ancestors = 100884679)